fhd_core/calibration/vis_calibrate_subroutine.pro

```diff
@@ -180,27 +190,69 @@ FUNCTION vis_calibrate_subroutine,vis_ptr,vis_model_ptr,vis_weight_ptr,obs,cal,p
                 BREAK
             ENDIF
             IF phase_fit_iter-i GT 0 THEN gain_new*=Abs(gain_old)*weight_invert(Abs(gain_new)) ;fit only phase at first
-;          IF (2.*phase_fit_iter-i GT 0) AND (phase_fit_iter-i LE 0) THEN $
-;            gain_new*=Mean(Abs(gain_new[where(gain_new)]))*weight_invert(Abs(gain_new)) ;then fit only average amplitude
-            gain_curr=(gain_new+gain_old)/2.
+            IF Keyword_Set(use_adaptive_gain) THEN $
+                conv_gain = calculate_adaptive_gain(conv_gain_list, convergence_list, i, base_gain=base_gain, final_convergence_estimate=0.) $
+                ELSE conv_gain = base_gain
+            gain_curr=(gain_new*conv_gain + gain_old*base_gain)/(base_gain + conv_gain)
             dgain=Abs(gain_curr)*weight_invert(Abs(gain_old))
             diverge_i=where(dgain LT Abs(gain_old)/2.,n_diverge)
             IF n_diverge GT 0 THEN gain_curr[diverge_i]=(gain_new[diverge_i]+gain_old[diverge_i]*2.)/3.
             IF nan_test(gain_curr) GT 0 THEN gain_curr[where(Finite(gain_curr,/nan))]=gain_old[where(Finite(gain_curr,/nan))]
             if ~keyword_set(no_ref_tile) then begin
               gain_curr*=Conj(gain_curr[ref_tile_use])/Abs(gain_curr[ref_tile_use])
             endif
-            conv_test[fii,i]=Max(Abs(gain_curr-gain_old)*weight_invert(Abs(gain_old)))
-            IF i GT phase_fit_iter THEN IF conv_test[fii,i] LE conv_thresh THEN BEGIN
-                n_converged += 1
-                BREAK
+            convergence_strict = Max(Abs(gain_curr-gain_old)*weight_invert(Abs(gain_old)))
+            convergence_loose = Mean(Abs(gain_curr-gain_old)*weight_invert(Abs(gain_old)))
+            convergence_list[i] = convergence_strict
+            conv_test[fii,i] = convergence_strict
+            IF i GT phase_fit_iter+divergence_history THEN BEGIN
```

Comment:
We should add another test here. If the solution diverges above `sqrt(conv_thresh)` then we should keep going, but save the last solution before that point. If the next solution from the next iteration is not better then we exit and use our saved solution, but if it does improve again we can keep going.